README.md
# go-endpoint

[![Build Status](https://github.com/beyondstorage/go-endpoint/workflows/Unit%20Test/badge.svg?branch=master)](https://github.com/beyondstorage/go-endpoint/actions?query=workflow%3A%22Unit+Test%22)
[![Go dev](https://pkg.go.dev/badge/github.com/beyondstorage/go-endpoint)](https://pkg.go.dev/github.com/beyondstorage/go-endpoint)
[![License](https://img.shields.io/badge/license-apache%20v2-blue.svg)](https://github.com/beyondstorage/go-endpoint/blob/master/LICENSE)
[![go storage dev](https://img.shields.io/matrix/go-endpoint:aos.dev.svg?server_fqdn=chat.aos.dev&label=%23go-endpoint%3Aaos.dev&logo=matrix)](https://matrix.to/#/#go-endpoint:aos.dev) <!-- Need update after matrix updated -->

Both human and machine readable  endpoint format.

## Format

```
<protocol>:<value>+
```

For example:

- File: `file:/var/cache/data`
- HTTP: `http:example.com:80`
- HTTPS: `https:example.com:443`

## Quick Start

```go
ep, err := endpoint.Parse("https:example.com")
if err != nil {
	log.Fatal("parse: ", err)
}

ep.Protocol() // -> "https"
ep.HTTPS() // -> "https://example.com", "example.com", 443
ep.HTTP() // -> panic
ep.File() // -> panic
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/)
and this project adheres to [Semantic Versioning](https://semver.org/).

## [Unreleased]

## [v1.1.0] - 2021-07-28

### Added

- Implementing RFC-8: Add TCP Protocol (#11)
- RFC-8: Add TCP protocol (#8)

## [v1.0.1] - 2021-06-15

### Fixed

- Fix protocol file that contains ":" (#4)
- Fix HTTP url generated incorrectly (#5)

## v1.0.0 - 2021-06-09

### Added

- Add Parse support
- Add http/https/file protocol support

[Unreleased]: https://github.com/beyondstorage/go-endpoint/compare/v1.0.1...HEAD
[v1.0.1]: https://github.com/beyondstorage/go-endpoint/compare/v1.0.0...v1.0.1


endpoint_test.go
package endpoint

import (
	"errors"
	"log"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestParse(t *testing.T) {
	cases := []struct {
		name  string
		cfg   string
		value Endpoint
		err   error
	}{
		{
			"invalid string",
			"abcx",
			Endpoint{},
			ErrUnsupportedProtocol,
		},
		{
			"normal http",
			"http:example.com:80",
			Endpoint{ProtocolHTTP, hostPort{"example.com", 80}},
			nil,
		},
		{
			"normal http without port",
			"http:example.com",
			Endpoint{ProtocolHTTP, hostPort{"example.com", 80}},
			nil,
		},
		{
			"wrong port number in http",
			"http:example.com:xxx",
			Endpoint{},
			ErrInvalidValue,
		},
		{
			"normal https",
			"https:example.com:443",
			Endpoint{ProtocolHTTPS, hostPort{"example.com", 443}},
			nil,
		},
		{
			"normal https without port",
			"https:example.com",
			Endpoint{ProtocolHTTPS, hostPort{"example.com", 443}},
			nil,
		},
		{
			"wrong port number in https",
			"https:example.com:xxx",
			Endpoint{},
			ErrInvalidValue,
		},
		{
			"not supported protocol",
			"notsupported:abc.com",
			Endpoint{},
			ErrUnsupportedProtocol,
		},
		{
			"normal file",
			"file:/root/data",
			Endpoint{ProtocolFile, "/root/data"},
			nil,
		},
		{
			"files contains `:`",
			"file:C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestStorage_Stat286526883\\001\\199446694",
			Endpoint{ProtocolFile, "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\TestStorage_Stat286526883\\001\\199446694"},
			nil,
		},
		{
			"normal tcp",
			"tcp:127.0.0.1:8000",
			Endpoint{ProtocolTCP, hostPort{"127.0.0.1", 8000}},
			nil,
		},
		{
			"wrong port number in tcp",
			"tcp:127.0.0.1:xxx",
			Endpoint{},
			ErrInvalidValue,
		},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			p, err := Parse(tt.cfg)
			if tt.err == nil {
				assert.Nil(t, err)
			} else {
				assert.True(t, errors.Is(err, tt.err))
			}
			assert.EqualValues(t, tt.value, p)
		})
	}
}

func TestNewFile(t *testing.T) {
	assert.Equal(t, Endpoint{ProtocolFile, "/example"}, NewFile("/example"))
}

func TestNewHTTP(t *testing.T) {
	assert.Equal(t,
		Endpoint{ProtocolHTTP, hostPort{"example.com", 8080}},
		NewHTTP("example.com", 8080),
	)
}

func TestNewHTTPS(t *testing.T) {
	assert.Equal(t,
		Endpoint{ProtocolHTTPS, hostPort{"example.com", 4433}},
		NewHTTPS("example.com", 4433),
	)
}

func TestNewTCP(t *testing.T) {
	assert.Equal(t,
		Endpoint{ProtocolTCP, hostPort{"127.0.0.1", 8000}},
		NewTCP("127.0.0.1", 8000),
	)
}

func TestEndpoint_Protocol(t *testing.T) {
	ep := NewFile("/test")

	assert.Equal(t, ProtocolFile, ep.Protocol())
}

func TestEndpoint_Protocol2(t *testing.T) {
	ep := NewTCP("127.0.0.1", 8000)

	assert.Equal(t, ProtocolTCP, ep.Protocol())
}

func TestEndpoint_String(t *testing.T) {
	cases := []struct {
		name     string
		value    Endpoint
		expected string
	}{
		{
			"file",
			Endpoint{ProtocolFile, "/test"},
			"file:/test",
		},
		{
			"http without port",
			Endpoint{ProtocolHTTP, hostPort{"example.com", 80}},
			"http:example.com:80",
		},
		{
			"http with port",
			Endpoint{ProtocolHTTP, hostPort{"example.com", 8080}},
			"http:example.com:8080",
		},
		{
			"https without port",
			Endpoint{ProtocolHTTPS, hostPort{"example.com", 443}},
			"https:example.com:443",
		},
		{
			"https with port",
			Endpoint{ProtocolHTTPS, hostPort{"example.com", 4433}},
			"https:example.com:4433",
		},
		{
			"tcp with port",
			Endpoint{ProtocolTCP, hostPort{"127.0.0.1", 8000}},
			"tcp:127.0.0.1:8000",
		},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.expected, tt.value.String())
		})
	}
}

func TestEndpoint(t *testing.T) {
	p := NewFile("/test")

	assert.Panics(t, func() {
		p.HTTP()
	})
	assert.Panics(t, func() {
		p.HTTPS()
	})

	assert.Panics(t, func() {
		p.TCP()
	})

	assert.Equal(t, "/test", p.File())
}

func TestEndpoint_HTTP(t *testing.T) {
	p := NewHTTP("example.com", 80)

	url, host, port := p.HTTP()
	assert.Equal(t, "http://example.com", url)
	assert.Equal(t, "example.com", host)
	assert.Equal(t, 80, port)

	p = NewHTTP("example.com", 8080)
	url, host, port = p.HTTP()
	assert.Equal(t, "http://example.com:8080", url)
	assert.Equal(t, "example.com", host)
	assert.Equal(t, 8080, port)
}

func TestEndpoint_HTTPS(t *testing.T) {
	p := NewHTTPS("example.com", 443)

	url, host, port := p.HTTPS()
	assert.Equal(t, "https://example.com", url)
	assert.Equal(t, "example.com", host)
	assert.Equal(t, 443, port)

	p = NewHTTPS("example.com", 4433)
	url, host, port = p.HTTPS()
	assert.Equal(t, "https://example.com:4433", url)
	assert.Equal(t, "example.com", host)
	assert.Equal(t, 4433, port)
}

func TestEndpoint_TCP(t *testing.T) {
	p := NewTCP("127.0.0.1", 8000)

	addr, host, port := p.TCP()
	assert.Equal(t, "127.0.0.1:8000", addr)
	assert.Equal(t, "127.0.0.1", host)
	assert.Equal(t, 8000, port)
}

func ExampleParse() {
	ep, err := Parse("http:example.com")
	if err != nil {
		log.Fatal(err)
	}

	switch ep.Protocol() {
	case ProtocolHTTP:
		url, host, port := ep.HTTP()
		log.Println("url: ", url)
		log.Println("host: ", host)
		log.Println("port: ", port)
	case ProtocolHTTPS:
		url, host, port := ep.HTTPS()
		log.Println("url: ", url)
		log.Println("host: ", host)
		log.Println("port: ", port)
	case ProtocolFile:
		path := ep.File()
		log.Println("path: ", path)
	case ProtocolTCP:
		addr, host, port := ep.TCP()
		log.Println("addr:", addr)
		log.Println("host:", host)
		log.Println("port", port)
	default:
		panic("unsupported protocol")
	}
}


endpoint.go
package endpoint

import (
	"fmt"
	"strconv"
	"strings"
)

const (
	// ProtocolHTTP is the http endpoint protocol.
	ProtocolHTTP = "http"
	// ProtocolHTTPS is the https endpoint protocol.
	ProtocolHTTPS = "https"
	// ProtocolFile is the file endpoint protocol
	ProtocolFile = "file"
	// ProtocolTCP is the tcp endpoint protocol
	ProtocolTCP = "tcp"
)

// Parse will parse config string to create a endpoint Endpoint.
func Parse(cfg string) (p Endpoint, err error) {
	s := strings.Split(cfg, ":")

	switch s[0] {
	case ProtocolHTTP:
		host, port, err := parseHostPort(s[1:])
		if err != nil {
			return Endpoint{}, &Error{"parse", ErrInvalidValue, s[0], []string{s[1], s[2]}}
		}
		if port == 0 {
			port = 80
		}
		return NewHTTP(host, port), nil
	case ProtocolHTTPS:
		host, port, err := parseHostPort(s[1:])
		if err != nil {
			return Endpoint{}, &Error{"parse", ErrInvalidValue, s[0], []string{s[1], s[2]}}
		}
		if port == 0 {
			port = 443
		}
		return NewHTTPS(host, port), nil
	case ProtocolFile:
		// Handle file paths that contains ":" (often happens on windows platform)
		//
		// See issue: https://github.com/beyondstorage/go-endpoint/issues/3
		path := strings.Join(s[1:], ":")
		return NewFile(path), nil
	case ProtocolTCP:
		//See issue: https://github.com/beyondstorage/go-endpoint/issues/7
		host, port, err := parseHostPort(s[1:])
		if err != nil {
			return Endpoint{}, &Error{"parse", ErrInvalidValue, s[0], []string{s[1], s[2]}}
		}
		return NewTCP(host, port), nil
	default:
		return Endpoint{}, &Error{"parse", ErrUnsupportedProtocol, s[0], nil}
	}
}

type hostPort struct {
	host string
	port int
}

func (hp hostPort) String() string {
	return fmt.Sprintf("%s:%d", hp.host, hp.port)
}

func parseHostPort(s []string) (host string, port int, err error) {
	if len(s) == 1 {
		return s[0], 0, nil
	}
	v, err := strconv.ParseInt(s[1], 10, 64)
	if err != nil {
		return "", 0, err
	}
	return s[0], int(v), nil
}

type Endpoint struct {
	protocol string
	args     interface{}
}

func NewHTTP(host string, port int) Endpoint {
	return Endpoint{
		protocol: ProtocolHTTP,
		args:     hostPort{host, port},
	}
}

func NewHTTPS(host string, port int) Endpoint {
	return Endpoint{
		protocol: ProtocolHTTPS,
		args:     hostPort{host, port},
	}
}

func NewFile(path string) Endpoint {
	return Endpoint{
		protocol: ProtocolFile,
		args:     path,
	}
}

func NewTCP(host string, port int) Endpoint {
	return Endpoint{
		protocol: ProtocolTCP,
		args:     hostPort{host, port},
	}
}

func (p Endpoint) Protocol() string {
	return p.protocol
}

func (p Endpoint) String() string {
	if p.args == nil {
		return p.protocol
	}
	return fmt.Sprintf("%s:%s", p.protocol, p.args)
}

func (p Endpoint) HTTP() (url, host string, port int) {
	if p.protocol != ProtocolHTTP {
		panic(Error{
			Op:       "http",
			Err:      ErrInvalidValue,
			Protocol: p.protocol,
			Values:   p.args,
		})
	}

	hp := p.args.(hostPort)
	if hp.port == 80 {
		return fmt.Sprintf("%s://%s", p.protocol, hp.host), hp.host, 80
	}
	return fmt.Sprintf("%s://%s:%d", p.protocol, hp.host, hp.port), hp.host, hp.port
}

func (p Endpoint) HTTPS() (url, host string, port int) {
	if p.protocol != ProtocolHTTPS {
		panic(Error{
			Op:       "https",
			Err:      ErrInvalidValue,
			Protocol: p.protocol,
			Values:   p.args,
		})
	}

	hp := p.args.(hostPort)
	if hp.port == 443 {
		return fmt.Sprintf("%s://%s", p.protocol, hp.host), hp.host, 443
	}
	return fmt.Sprintf("%s://%s:%d", p.protocol, hp.host, hp.port), hp.host, hp.port
}

func (p Endpoint) File() (path string) {
	if p.protocol != ProtocolFile {
		panic(Error{
			Op:       "file",
			Err:      ErrInvalidValue,
			Protocol: p.protocol,
			Values:   p.args,
		})
	}

	return p.args.(string)
}

func (p Endpoint) TCP() (addr, host string, port int) {
	if p.protocol != ProtocolTCP {
		panic(Error{
			Op:       "tcp",
			Err:      ErrInvalidValue,
			Protocol: p.protocol,
			Values:   p.args,
		})
	}
	hp := p.args.(hostPort)
	return fmt.Sprintf("%s:%s:%d", p.protocol, hp.host, hp.port), hp.host, hp.port
}
